### PR TITLE
[16.0][ADD] stock_grn

### PR DESCRIPTION
**Pull request by glitchov**
Module to link a goods received note to a receipt

### CODE REVIEW

File: stock_grn/models/stock_grn.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
+# Author: Jacques-Etienne Baudoux <je@bcim.be>
+# Copyright (C) 2015-TODAY BCIM <http://www.bcim.be>
+#                          ACSONE SA/NV
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+
+from odoo import api, fields, models
+
+
+class GRNType(models.Model):
```
Comment (rousseldenis) [inline stock_grn/models/stock_grn.py]: Please split one model per file

File: stock_grn/models/stock_picking.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
+# Author: Jacques-Etienne Baudoux <je@bcim.be>
+# Copyright (C) 2015-TODAY BCIM <http://www.bcim.be>
+#                          ACSONE SA/NV
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+
+
+from odoo import fields, models
+
+
+class StockPicking(models.Model):
```
Comment (rousseldenis) [inline stock_grn/models/stock_picking.py:6]: Same as above

File: stock_grn/models/stock_picking.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
+# Author: Jacques-Etienne Baudoux <je@bcim.be>
+# Copyright (C) 2015-TODAY BCIM <http://www.bcim.be>
```
Comment (rousseldenis) [inline stock_grn/models/stock_picking.py]:
```suggestion
# Copyright (C) 2015-2022 BCIM <http://www.bcim.be>
```

The code is not changing magically everyday :-)

File: stock_grn/readme/USAGE.rst [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+============
+
+To install this module, you just need to select the module and insure yourself dependencies are available.
+
+Usage
+=====
+
+To use this module, you need to :
+- create a GRN
+- make a reception and attach the GRN
```
Comment (rousseldenis) [inline stock_grn/readme/USAGE.rst]: Functionnally tested. Impossible to attach a GRN on picking level.

File: stock_grn/security/ir_rule.xml [hunk 5]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+  <record model="ir.rule" id="grn_rule">
+    <field name="name">grn company rule</field>
+    <field name="model_id" ref="model_stock_grn" />
+    <field name="global" eval="True" />
+    <field
+            name="domain_force"
+        >['|',('company_id','=',False),('company_id','=',user.company_id.id)]</field>
```
Comment (rousseldenis) [inline stock_grn/security/ir_rule.xml]:
```suggestion
        >['|',('company_id','=',False),('company_id','in',company_ids)]</field>
```

File: stock_grn/views/stock_picking_views.xml [hunk 6]
```diff
@@ -0,0 +1,68 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+  <record id="view_picking_withgrn_search" model="ir.ui.view">
+    <field name="name">stock.picking.grn.search</field>
+    <field name="model">stock.picking</field>
+    <field name="inherit_id" ref="stock.view_picking_internal_search" />
+    <field name="arch" type="xml">
+      <filter name="late" position="after">
+        <filter name="grn" string="Received" domain="[('grn_id', '!=', False)]" />
```
Comment (rousseldenis) [inline stock_grn/views/stock_picking_views.xml]: Maybe something more like :

```suggestion
        <filter name="grn" string="Received (GRN)" domain="[('grn_id', '!=', False)]" />
```

No ?

**Review by rousseldenis — CHANGES_REQUESTED** [review]

**Comment by rousseldenis** [comment]
@glitchov

File: stock_grn/security/ir_rule.xml (hunk 5 above)
Comment (rousseldenis) [reply stock_grn/security/ir_rule.xml]: @glitchov I suggested 'in' :sweat_smile:

Comment (glitchov) [reply stock_grn/security/ir_rule.xml]: Corrected in between ;-)

**Review by lmignon — APPROVED** [review]
LGTM (Code review)

File: stock_grn/readme/USAGE.rst [hunk 7]
```diff
@@ -0,0 +1,11 @@
+Installation
+============
```
Comment (rousseldenis) [inline stock_grn/readme/USAGE.rst]: You should not use titles if not needed.

Split this with an INSTALL.rst (with just the 'To install this module ...' sentence)

File: stock_grn/readme/USAGE.rst [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+
+To install this module, you just need to select the module and insure yourself dependencies are available.
+
+Usage
+=====
```
Comment (rousseldenis) [inline stock_grn/readme/USAGE.rst]: This title will be generated automatically

File: stock_grn/models/stock_grn_type.py [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+from odoo import fields, models
+
+
+class GRNType(models.Model):
+    """GRN Type"""
+
+    _name = "stock.grn.type"
+    _description = "Type of goods received note"
+
+    name = fields.Char(string="Type", required=True)
```
Comment (FrancoMaxime) [inline stock_grn/models/stock_grn_type.py:13]: For this kind of model that has only a name field, i usually add a SQL constraint to ensure the name is unique. Does it have sense here?

Comment (glitchov) [reply stock_grn/models/stock_grn_type.py:13]: @jbaudoux  What do you think of  this ?

File: stock_grn/README.rst [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+Contributors
+------------
+
+* Jacques-Etienne Baudoux <je@bcim.be> (BCIM sprl)
+* Hughes Damry <hughes.damry@acsone.eu>
+
+Maintainer
+----------
+
+.. image:: http://www.qansee.com/logo.png
```
Comment (jbaudoux) [inline stock_grn/README.rst]: should be regenerated to have OCA

File: stock_grn/__manifest__.py [hunk 11]
```diff
@@ -0,0 +1,25 @@
+# Author: Jacques-Etienne Baudoux <je@bcim.be>
+# Copyright (C) 2015-TODAY BCIM <http://www.bcim.be>
+#                          ACSONE SA/NV
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
```
Comment (jbaudoux) [inline stock_grn/__manifest__.py]:
```suggestion
# Copyright 2015 Jacques-Etienne Baudoux (BCIM) <je@bcim.be>
# Copyright 2020 ACSONE SA/NV
# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
```

File: stock_grn/models/stock_grn.py [hunk 12]
```diff
@@ -0,0 +1,57 @@
+# Author: Jacques-Etienne Baudoux <je@bcim.be>
+# Copyright (C) 2015-2022 BCIM <http://www.bcim.be>
+#                         ACSONE SA/NV
```
Comment (jbaudoux) [inline stock_grn/models/stock_grn.py]:
```suggestion
# Copyright 2015 Jacques-Etienne Baudoux (BCIM) <je@bcim.be>
```

File: stock_grn/models/stock_grn.py [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+
+from odoo import api, fields, models
+
+
+class GRN(models.Model):
```
Comment (jbaudoux) [inline stock_grn/models/stock_grn.py]:
```suggestion
class StockGrn(models.Model):
```

File: stock_grn/models/stock_grn_type.py [hunk 14]
```diff
@@ -0,0 +1,15 @@
+# Author: Jacques-Etienne Baudoux <je@bcim.be>
+# Copyright (C) 2015-2022 BCIM <http://www.bcim.be>
+#                         ACSONE SA/NV
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
```
Comment (jbaudoux) [inline stock_grn/models/stock_grn_type.py]:
```suggestion
# Copyright 2015 Jacques-Etienne Baudoux (BCIM) <je@bcim.be>
# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
```

File: stock_grn/models/stock_grn_type.py (hunk 9 above)
Comment (jbaudoux) [reply stock_grn/models/stock_grn_type.py:13]: Would be a good addition

File: stock_grn/models/stock_picking.py [hunk 15]
```diff
@@ -0,0 +1,27 @@
+# Author: Jacques-Etienne Baudoux <je@bcim.be>
+# Copyright (C) 2015-2022 BCIM <http://www.bcim.be>
+#                         ACSONE SA/NV
```
Comment (jbaudoux) [inline stock_grn/models/stock_picking.py]:
```suggestion
# Copyright 2015 Jacques-Etienne Baudoux (BCIM) <je@bcim.be>
```

File: stock_grn/models/stock_picking_type.py [hunk 16]
```diff
@@ -0,0 +1,34 @@
+# Author: Jacques-Etienne Baudoux <je@bcim.be>
+# Copyright (C) 2015-2022 BCIM <http://www.bcim.be>
+#                         ACSONE SA/NV
```
Comment (jbaudoux) [inline stock_grn/models/stock_picking_type.py]:
```suggestion
# Copyright 2015 Jacques-Etienne Baudoux (BCIM) <je@bcim.be>
```

File: stock_grn/readme/INSTALL.rst [hunk 17]
```diff
@@ -0,0 +1 @@
+To install this module, you just need to select the module and insure yourself dependencies are available.
```
Comment (jbaudoux) [inline stock_grn/readme/INSTALL.rst]: This file can be dropped

**Review by jbaudoux — CHANGES_REQUESTED** [review]
Can you cleanup the copyrights?
Otherwise, looks good

**Comment by lmignon** [comment]
@jbaudoux CAn you update your review? Changes done.

**Review by jbaudoux — APPROVED** [review]
Thanks

File: stock_grn/security/ir.model.access.csv [hunk 18]
```diff
@@ -0,0 +1,5 @@
+id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
+user_001,user_001,model_stock_grn_type,stock.group_stock_user,1,0,0,0
```
Comment (rousseldenis) [inline stock_grn/security/ir.model.access.csv:2]: Maybe better id and name ?

**Review by rousseldenis — APPROVED** [review]
Minor comment

**Review by FrancoMaxime — APPROVED** [review]
LGTM (code review)

**Pull request merged**